draft-specifications/Model.md

```diff
@@ -0,0 +1,326 @@
+> IEEE 754, except that OCapN considers all IEEE 754 NaNs as equivalent, i.e., as
+> jointly representing a single abstract NaN value. Thus, any concrete NaN representation
+> may validly round trip even if it results in a different concrete representation.
+> However, we encourage concrete representations to use a canonical NaN representation.
+>
+> Concretely, the canonical NaN is `0x7ff8\_0000\_0000\_0000`, though this is not
+> a concern of the abstract syntax and data model.
+>
+> All real and finite double precision floating point numbers participate in the
+> JSON subset of OCapN.
+> We expect OCapN-compatible JSON codecs, including the JavaScript `JSON` codec,
+> to round-trip all numbers except NaNs, infinities, and negative zero, but all
+> other numbers expressible with an IEEE 754 double-precision float to survive a
+> round trip without loss of precision.
+> We also do not expect integers expressed with higher precision in JSON to
+> survive a round-trip through an OCapN-compatible JSON codec.
+>
+> Consensus on preserving -0:
+> - As of May 16, @erights insisted that -0 round trip to 0 https://github.com/ocapn/ocapn/issues/5#issuecomment-1550020450
+> - On May 23, @zenhack proposed preserving -0 https://github.com/ocapn/ocapn/issues/5#issuecomment-1560116857
+> - We converged out of band and Agoric has committed to preserving -0 https://github.com/endojs/endo/issues/1602
+
+## String (JSON\*)
+
+A string of Unicode code points but no lone/unpaired surrogates.
+
+> - **Guile**: `""`
+> - **JavaScript**: `''`
+> - **Python**: `''`
+>
+> \*Strings participate in the JSON subset of OCapN except any strings that
+> contain unpaired/lone surrogates.
+>
+> OCapN cannot send or receive strings that come from a UTF-16 language
+> representation (such as a JavaScript string) that contains lone/unpaired
+> surrogates.
+> Notes: [Jan 2024 meeting
+> notes](https://cryptpad.fr/code/#/2/code/edit/Ng7kXl9rCF5qnTQX60UCpDHE/)
+> record that we agreed that strings can only be well-formed Unicode, i.e.,
+> cannot contain unpaired surrogates.
+> For JavaScript, if a string does not pass [the `isWellFormed`
+> predicate](https://github.com/tc39/proposal-is-usv-string), then it is not a
+> Passable string.
+>
+> Agoric has agreed to tentatively omit lone surrogates, pending vetting of the
+> performance of the implementation in their CapTP.
+> https://github.com/ocapn/ocapn/issues/47
+>
+> Agoric previously held the position that all valid strings that JavaScript
+> can parse with `JSON.parse` should round-trip.
+> https://github.com/ocapn/ocapn/issues/5#issuecomment-1550370096,
+> https://github.com/ocapn/ocapn/issues/5#issuecomment-1550481111
+>
+> Spritely has consistently preferred a representation that can be captured
+> with UTF-8, which cannot transit lone/unpaired surrogates.
+> (October 10, 2023)
+
+## ByteArray
+
+An array of bytes.
+
+> - **Guile**: `#vu8()`
+> - **JavaScript**: `new ArrayBuffer()`
+> - **Python**: `b''`
```

Comment:
The Guile and Python examples show literals but the JS one is a mutable thing. In Guile, the equivalent to `new ArrayBuffer()` to make a mutable bytevector would be `(make-bytevector len)`.

Reply:
@erights is proposing immutable ArrayBuffers this very week, so I anticipate updating this, but this is close enough for now.